src/pool.rs

```diff
@@ -603,19 +614,53 @@ impl ConnectionPool {
     /// Get a connection from the pool.
     pub async fn get(
         &self,
-        shard: usize,               // shard number
+        shard: Option<usize>,       // shard number
         role: Option<Role>,         // primary or replica
         client_stats: &ClientStats, // client id
     ) -> Result<(PooledConnection<'_, ServerPool>, Address), Error> {
-        let mut candidates: Vec<&Address> = self.addresses[shard]
-            .iter()
-            .filter(|address| address.role == role)
-            .collect();
+        let mut effective_shard_id: Option<usize> = shard;
+
+        // The base, unsharded case
+        if self.shards() == 1 {
+            effective_shard_id = Some(0);
+        }
+
+        let mut sort_by_error_count = false;
+        let mut candidates: Vec<_> = match effective_shard_id {
+            Some(shard_id) => self.addresses[shard_id].iter().collect(),
+            None => match self.settings.no_shard_specified_behavior {
+                NoShardSpecifiedHandling::Random => self.addresses.iter().flatten().collect(),
+                NoShardSpecifiedHandling::RandomHealthy => {
+                    sort_by_error_count = true;
+                    self.addresses.iter().flatten().collect()
+                }
+                NoShardSpecifiedHandling::Shard(shard) => {
+                    if shard >= self.shards() {
+                        return Err(Error::InvalidShardId(shard));
+                    } else {
+                        self.addresses[shard].iter().collect()
```

Comment:
Should also range check on addresses here. It panics with the comment based sharding if an invalid shard is passed

Reply:
Do you mean like -1?

Reply:
Mostly for when shard is higher than the addresses list. With a manually built query and the comment based sharding passing in a shard that's too high results in a panic instead of a client error.